library/alloc/src/collections/btree/node.rs

```diff
@@ -1294,32 +1296,34 @@ impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Internal>, marker::
 
 impl<'a, K, V> NodeRef<marker::Mut<'a>, K, V, marker::LeafOrInternal> {
     /// Chooses a balancing context involving the node as a child, thus between
-    /// the KV immediately to the left or to the right in the parent node.
-    /// Returns an `Err` if there is no parent.
+    /// balancing around the parent's KV immediately to the left or to the right
+    /// of the child. Returns an `Err` if there is no or an empty parent.
     ///
-    /// This method optimizes for a node that has fewer elements than its left
-    /// and right siblings, if they exist, by preferring the left parent KV.
-    /// Merging with the left sibling is faster, since we only need to move
+    /// Prefers the left side, to be optimal if the given node is somehow
+    /// underfull, meaning here only that it has fewer elements than its left
+    /// sibling and than its right sibling, if they exist. In that case,
+    /// merging with the left sibling is faster, since we only need to move
     /// the node's N elements, instead of shifting them to the right and moving
     /// more than N elements in front. Stealing from the left sibling is also
     /// typically faster, since we only need to shift the node's N elements to
     /// the right, instead of shifting at least N of the sibling's elements to
     /// the left.
     pub fn choose_parent_kv(self) -> Result<LeftOrRight<BalancingContext<'a, K, V>>, Self> {
-        match unsafe { ptr::read(&self) }.ascend() {
-            Ok(parent) => match parent.left_kv() {
+        let child = unsafe { ptr::read(&self) };
+        match self.ascend() {
+            Ok(parent_edge) => match parent_edge.left_kv() {
                 Ok(left_parent_kv) => Ok(LeftOrRight::Left(BalancingContext {
                     parent: unsafe { ptr::read(&left_parent_kv) },
                     left_child: left_parent_kv.left_edge().descend(),
-                    right_child: self,
+                    right_child: child,
                 })),
-                Err(parent) => match parent.right_kv() {
+                Err(parent_edge) => match parent_edge.right_kv() {
                     Ok(right_parent_kv) => Ok(LeftOrRight::Right(BalancingContext {
                         parent: unsafe { ptr::read(&right_parent_kv) },
-                        left_child: self,
+                        left_child: child,
                         right_child: right_parent_kv.right_edge().descend(),
                     })),
-                    Err(_) => unreachable!("empty non-root node"),
+                    Err(parent_edge) => Err(parent_edge.descend()),
```

Comment:
Am I wrong that `parent_edge.descend() == self` here? AFAICT, if we're in this branch that should be the case, but maybe I'm overlooking something?

Regardless it seems like we *should* error out on such an invalid state (at least a debug assert seems reasonable); do you think otherwise? I guess this is used during balancing, and we may have temporarily invalid nodes in that context.

Reply:
> Am I wrong that `parent_edge.descend() == self` here?

Indeed. Well, `child` at the moment, but now I don't see the point taking that child snapshot up front.

> Regardless it seems like we _should_ error out on such an invalid state (at least a debug assert seems reasonable); do you think otherwise?

I'm balancing on one foot on the otherwise side. The node module doesn't list emptiness (or underfullness) as an invariant, and we can't easily impose it. None of the node methods list it as a requirement, though others  like `choose_parent_kv` may forget to mention it, and for instance `key_at` implies the node cannot be empty. We do have empty parent nodes during map algorithms, and at times not handling empty parents the same as absent parents caused trouble, though now I'm pretty sure the algorithms guarantee that the map layer can never invoke `choose_parent_kv` on the child of an empty node.

However the source of this PR is what it didn't fix: my annoyance with the type erasure and the `cast_to_leaf_unchecked` up in remove.rs. Let me try a little harder to work that out.